File: leetcode-cn/0011.0_Container_With_Most_Water.py
'''
Brute Force
Time: O(N ^ 2)
Space:O(1)

执行结果：
超出时间限制
显示详情
最后执行的输入：
[1454,2249,4922,5151,7349,1022,8655,8124,2157,2303,5213,3217,437...
'''

class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        maximum = 0
        size = len(height)
        for i in range(size - 1):
            for j in range(i + 1, size):
                maximum = max(maximum, min(height[i], height[j]) * (j - i))
        return maximum

File: cpp/0011.0_Container_With_Most_Water.cpp
/*
执行用时：72 ms, 在所有 C++ 提交中击败了68.89% 的用户
内存消耗：57.5 MB, 在所有 C++ 提交中击败了86.44% 的用户
通过测试用例：60 / 60
*/

class Solution {
public:
    int maxArea(vector<int>& height) {
        int ans = 0;
        int i = 0, j = height.size() - 1;
        while (i < j) {
            ans = max(ans, min(height[i], height[j]) * (j - i));
            if (height[i] < height[j]) {
                ++i;
            } else {
                --j;
            }
        }
        return ans;
    }
};
